pylint/checkers/typecheck.py

```diff
@@ -1035,6 +1053,25 @@ def visit_assign(self, node):
             else:
                 self.add_message("assignment-from-none", node=node)
 
+    def _check_dundername_is_string(self, node):
+        """
+        Check a string is assigned to self.__name__
+        """
+
+        # Check the left hand side of the assignment is <something>.__name__
+        lhs = list(node.get_children())[0]
+        if not isinstance(lhs, astroid.node_classes.AssignAttr):
+            return
+        if not lhs.attrname == "__name__":
+            return
+
+        # If the right hand side is not a string
+        rhs = list(node.get_children())[1]
+        if not (isinstance(rhs, astroid.node_classes.Const)
```

Comment:
Unfortunately it's not quite that simple, we need to infer the value on the right hand side to figure out if it is a string or not. To speed it up, we can still use the `Const` guard, so something along these lines should do the trick:
```
if isinstance(rhs, astroid.Const) and isinstance(rhs.value, str):
     return
inferred = utils.safe_infer(rhs)
if not inferred:
    return
if not (isinstance(inferred, astroid.Const) and isinstance(inferred.value, str)):
    # emit the message
```

Reply:
Thank you! Should be fixed in https://github.com/PyCQA/pylint/pull/3271/commits/abbcafab1f23b72cfc37af10861d3ded58dd7a6d